### DISCUSSION TITLE
How to control the amplitude and the frequency of the generator?

### DISCUSSION THREAD

**Discussion by wimverlindensa** [pschatzmann/arduino-audio-tools#164] [Q&A] [answered]
I would like to generate a sinus tone which works with below code :
(https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-generator-i2s/README.md)

But how can I control the tone in amplitude and in frequency? Is that somewhere explained?
I would like to control start/stop/freq/ampl by a webinterface and buttons, but not find the methods to do that.

What would be the max frequency possible?

When I   commented //AudioLogger::instance().begin(Serial, AudioLogger::Info);  the sound was perfect, but uncommented I got a very disturbed sound. ( using PCM5102 module)

Thanks for feedback!

**Comment by pschatzmann** [comment, marked as answer]
I suggest that you check the [class documentation](https://pschatzmann.github.io/arduino-audio-tools/html/annotated.html) and look up SineWaveGenerator. If you plan to change the parameters of a running tone I suggest that you use the SineFromTable instead as  documented. 

Instead of changing the amplitude you might also consider to use the [Volume Control](https://github.com/pschatzmann/arduino-audio-tools/wiki/Volume-Control) as described in the Wiki

You can comment out the log statement or set it to Warning as it is recommended in the Readme.

Starting/Stopping can be done in many ways:

- by setting to volume to 0 and back to the original value
- by setting to frequency to 0 and back to the original value
- with a flag that activates/deactivates the copy in the loop

For the max frequency I am not sure. I would guess it is half the sampling rate ? Though I would expect that will not generate any nice sine wave and I am not sure if a regular speaker would be able to go above 20'000Hz
